src/main.rs

```diff
@@ -125,13 +177,101 @@ fn main() {
     }
 
 
-    println!("Enter `exit` to exit.");
+    println!("Enter `exit` (or Control-D) to exit.");
     let mut input = String::new();
-    while input.trim() != "exit" {
+    loop {
         input.clear();
         print!("> ");
         io::stdout().flush().unwrap(); // https://github.com/rust-lang/rust/issues/23818
-        io::stdin().read_line(&mut input).ok();
+        // Note: read_line includes the newline character.
+        if let Ok(n) = io::stdin().read_line(&mut input) {
+            if n == 0 {
+                // End-of-file (either Control-D or we were redirected).
+                break;
+            }
+        } else {
+            println!("Error reading stdin");
+            break;
+        }
+
+        let tokens:Vec<&str> = input.split_whitespace().collect();
+        if tokens.len() == 0 {
+            continue;
+        }
+        match tokens[0] {
+            "args" => {
+                println!("args = {:?}", tokens);
+            }
+            "exit" => {
+                break;
+            }
+            "controllers" => {
+                let ref controllers = program.state.lock().unwrap().controllers;
+                for controller in controllers {
+                    print_controller(&controller);
```

Comment:
if you make this the Display implementation, you'll be able to use `println!("{}", controller)` :)